useful_resources/sample_code/opt_outs_everyaction.py

```diff
@@ -0,0 +1,163 @@
+
+        if isinstance(r, dict):
+            # Response is only a dict upon success
+            success_log.append({
+                "vanid": r.get('vanId'),
+                "phone": phone,
+                "committeeid": committeeid,
+                "applied_at": applied_at
+            })
+
+    # If we get an HTTP Error we log it
+    # Usually these errors are due to a vanid being deleted or merged
+    except requests.exceptions.HTTPError as e:
+        r = e
+        error_log.append({
+            "vanid": vanid,
+            "phone": phone,
+            "committeeid": committeeid,
+            "errored_at": applied_at,
+            "error": r
+        })
+
+    # If we get a connection error we wait a bit and try again.
+    except requests.exceptions.ConnectionError as c:
+        logger.info("Got disconnected, waiting and trying again")
+
+        while attempts_left > 0:
+            attempts_left -= 1
+
+            # Wait 10 seconds, then try again
+            time.sleep(10)
+            attempt_optout(ea, row, attempts_left)
+
+        else:
+            r = str(c)[:999]
+            return r
+
+            error_log.append({
+                "vanid": vanid,
+                "phone": phone,
+                "committeeid": committeeid,
+                "errored_at": applied_at,
+                "error": r
+            })
+            error_tbl = Table(error_log)
+            if error_tbl.num_rows() > 0:
+                error_tbl.to_redshift(
+                    ERROR_TABLE,
+                    if_exists='append',
+                    alter_table=True
+                )
+            raise Exception(f"Connection Error {r}")
+
+    return r
+
+
+def main():
+    # Creating empty lists where we'll log successes and errors
+    success_log = []
+    error_log = []
+
+    # Get the opt out data
+    s = rs.query(f"select * from {OPT_OUT_TABLE}")
+
+    # Turn the API keys into a list
+    API_KEYS = list(ast.literal_eval(API_KEYS_STR))
+
+    # Loop through each API key to update phones in each committee
+    for k in API_KEYS:
+
+        api = k['api_key']
+        committeeid = k['committee_id']
+        committee = k['committee']  # This is the committee name
+        ea = VAN(db='EveryAction', api_key=api)
+
+        logger.info(f"Working on opt outs in {committee} committee...")
+
+        update = s.select_rows(lambda row: str(row.committeeid) == committeeid)
+
+        logger.info(f"Found {update.num_rows} phones to opt out in {committee} committee...")
+
+        # Now we actually update the records
+
+        if update.num_rows > 0:
+
+            for u in update:
+
+                applied_at = str(datetime.now()).split(".")[0]
+                attempt_optout(ea, u, applied_at, committeeid, success_log, error_log)
+
+    # Now we log results
+    logger.info(f"There were {len(success_log)} successes and {len(error_log)} errors.")
+
+    if len(success_log) > 0:
+        success_t = Table(success_log)
```

Comment:
Might as well make this variable `success_table`, for readability, though I'm sure folks could guess.  Although I'm possibly a bit confused about what's going on here.

My goal here is that the variables be named in ways that clearly and accurately describe what they are, to make it as easy as possible for folks to read the script. So please feel free to try other names if the ones I'm suggesting are wrong or a bad fit.

Whatever we decide, the naming pattern should be the same for the error table too.

Reply:
I ended up renaming this and `error_t` to `success_parsonstable` and `error_parsonstable` because we already have variables called `SUCCESS_TABLE` and `ERROR_TABLE` which are the Redshift table names!